File: GameUpdate/GameUpdate/EditorCore/Common/CustomThread.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;

// 返回值表示是否继续运行该线程
public delegate bool CustomThreadCallback();

public class CustomThread : EditorBase
{
	protected bool mRunning;
	protected bool mFinish;
	protected bool mPause;
	protected CustomThreadCallback mCallback;
	protected Thread mThread;
	protected ThreadTimeLock mTimeLock;
	protected string mName;
	protected bool mIsBackground;		// 是否为后台线程,如果是后台线程,则在应用程序关闭时,子线程会自动强制关闭
	public CustomThread(string name)
	{
		mName = name;
		mFinish = true;
		mRunning = false;
		mCallback = null;
		mThread = null;
		mTimeLock = null;
		mPause = false;
		mIsBackground = true;
	}
	public void destroy()
	{
		stop();
	}
	public void setBackground(bool background)
	{
		mIsBackground = background;
		if(mThread != null)
		{
			mThread.IsBackground = mIsBackground;
		}
	}
	public void start(CustomThreadCallback callback, int frameTimeMS = 15)
	{
		logConsole("准备启动线程 : " + mName);
		if (mThread != null)
		{
			return;
		}
		mTimeLock = new ThreadTimeLock(frameTimeMS);
		mRunning = true;
		mCallback = callback;
		mThread = new Thread(run);
		mThread.Name = mName;
		mThread.Start();
		mThread.IsBackground = mIsBackground;
		logConsole("线程启动成功 : " + mName);
	}
	public void pause(bool isPause)
	{
		mPause = isPause;
	}
	public void stop()
	{
		logConsole("准备退出线程 : " + mName);
		mRunning = false;
		while (!mIsBackground && !mFinish) { }
		if (mThread != null && !mIsBackground)
		{
			mThread.Abort();
		}
		mThread = null;
		mCallback = null;
		mTimeLock = null;
		mPause = false;
		logConsole("线程退出完成! 线程名 : " + mName);
	}
	protected void run()
	{
		mFinish = false;
		while (mRunning)
		{
			mTimeLock.update();
			try
			{
				if(mPause)
				{
					continue;
				}
				if(!mCallback())
				{
					break;
				}
			}
			catch (Exception e)
			{
				logConsole("捕获线程异常! 线程名 : " + mName + ", " + e.Message + ", " + e.StackTrace);
			}
		}
		mFinish = true;
	}
}

File: Engine/Thread/CustomThread.cpp
#include "CustomThread.h"
#include "Utility.h"
#include "txMemoryTrace.h"
#include "TimeLock.h"

CustomThread::CustomThread(const std::string& name)
{
	mName = name;
	mFinish = true;
	mRunning = false;
	mCallback = NULL;
	mThread = NULL;
	mTimeLock = NULL;
	mPause = false;
	mIsBackground = true;
}

void CustomThread::destroy()
{
	stop();
}

void CustomThread::start(CustomThreadCallback callback, void* args, const int& frameTimeMS)
{
	ENGINE_INFO("准备启动线程 : %s", mName.c_str());
	if (mThread != NULL)
	{
		return;
	}
	mTimeLock = TRACE_NEW(TimeLock, mTimeLock, frameTimeMS);
	mTimeLock->setForceSleep(5);
	mRunning = true;
	mCallback = callback;
	mArgs = args;
	CREATE_THREAD(mThread, run, this);
	ENGINE_INFO("线程启动成功 : %s", mName.c_str());
}

void CustomThread::stop()
{
	ENGINE_INFO("准备退出线程 : %s", mName.c_str());
	mRunning = false;
	mPause = false;
	while (!mIsBackground && !mFinish) {}
	CLOSE_THREAD(mThread);
	mCallback = NULL;
	TRACE_DELETE(mTimeLock);
	ENGINE_INFO("线程退出完成! 线程名 : %s", mName.c_str());
}

void CustomThread::updateThread()
{
	mFinish = false;
	while (mRunning)
	{
		mTimeLock->update();
		try
		{
			if (mPause)
			{
				continue;
			}
			if (!mCallback(mArgs))
			{
				break;
			}
		}
		catch (...)
		{
			ENGINE_INFO("捕获线程异常! 线程名 : %s", mName.c_str());
		}
	}
	mFinish = true;
}